### PR TITLE
init backend db, api frame

### PR DESCRIPTION
**Pull request by baidang201**

### CODE REVIEW

File: backend/proto/message.proto [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package  launchpadpackage;
+syntax = "proto3";
+
+message Worker {
+  string stashAccount = "";
+  string controllerAccount = "";
+  string payout = "";
+  string status = "";
+  bool   stakeEnough = false;
+  string accumulatedStake = "0";// bignum
```
Comment (h4x3rotab) [inline backend/proto/message.proto:10]:
```suggestion
  string accumulatedStake = "0";  // bignum
```

Code style

File: backend/proto/message.proto [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package  launchpadpackage;
+syntax = "proto3";
+
+message Worker {
```
Comment (h4x3rotab) [inline backend/proto/message.proto]: Proto definition should be like `Type snake_name = <field_id>`. E.g.:

```
string stash_account = 1;
string controller_account = 2;
// ...
float32 commission = 10;
```

And usually in the proto libraries the `snake_cased` names will be converted to `camelCased`.

In proto3 you cannot specify a default value to a field.

https://developers.google.com/protocol-buffers/docs/proto3#default

Comment (baidang201) [reply backend/proto/message.proto]: fixed

File: backend/models/realtimeRoundInfo.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import mongoose from  "../db/mogoose"
+
+const Worker = new mongoose.Schema({ 
+  stashAccount: {tyep: String},
```
Comment (h4x3rotab) [inline backend/models/realtimeRoundInfo.js]: type?

File: proto/message.proto [hunk 4]
```diff
@@ -0,0 +1,98 @@
+syntax = "proto3";
+package  launchpadpackage;
+
+message Worker {
+  string stashAccount = 1;
```
Comment (h4x3rotab) [inline proto/message.proto]: Could you check if protobufjs automatically converts snake_case to camelCase?

The standard of protobuf is always snake_case

File: proto/message.proto [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  string controllerAccount = 2;
+  string payout = 3;
+  string status = 4;
+  bool   stakeEnough = 5;
+  string accumulatedStake = 6;// bignum
```
Comment (h4x3rotab) [inline proto/message.proto]: 2 spaces before "//", everywhere

File: backend/models/realtimeRoundInfo.js (hunk 3 above)
Comment (baidang201) [reply backend/models/realtimeRoundInfo.js]: done

File: proto/message.proto (hunk 5 above)
Comment (baidang201) [reply proto/message.proto]: done

File: proto/message.proto (hunk 4 above)
Comment (baidang201) [reply proto/message.proto]: done

File: backend/servers/history.js [hunk 6]
```diff
@@ -0,0 +1,382 @@
+import pQueue from 'p-queue'
+import { ApiPromise, WsProvider } from '@polkadot/api'
+import Demical from 'decimal.js'
```
Comment (jasl) [inline backend/servers/history.js]: `Decimal`

**Review by jasl — CHANGES_REQUESTED** [review]

File: backend/servers/history.js [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+  }
+
+  processRoundAt = async (header, api) => {
+    const blockHash = header.hash
+
+    const roundInfo = (await api.query.phalaModule.round.at(blockHash)) || new BN('0')
+    const roundNumber = roundInfo.round.toNumber()
+    const number = header.number.toNumber()
+    const accumulatedFire2 = (await api.query.phalaModule.accumulatedFire2.at(blockHash)) || new BN('0')
+    const accumulatedFire2Demical = new Demical(accumulatedFire2.toString())
```
Comment (jasl) [inline backend/servers/history.js]: Decimal

File: backend/servers/history.js [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+        .map(async k => {
+          const account = k.args[0].toString()
+          const value = await api.rpc.state.getStorage(k, blockHash)
+          const fire2 = value.toString()
+          payoutAccounts[account] = {
+            ...payoutAccounts[account],
+            account,
+            fire2,
+            fire2Human: value.toHuman().replace(/PHA$/, '').replace(' ', ''),
+            prizeRatio: new Demical(fire2).div(accumulatedFire2Demical).toNumber(),
```
Comment (jasl) [inline backend/servers/history.js]: Decimal

**Comment by jasl** [comment]
Demical should be Decimal

**Review by jasl — CHANGES_REQUESTED** [review]

File: backend/index.js [hunk 9]
```diff
@@ -1 +1,89 @@
...
+  'global_statistics_request'     : function(obj) {  const buffer = getGlobalStatistics();  return buffer },
+  'avg_stake_request'             : function(obj) {  const buffer = getAvgStake();  return buffer},
+  'stake_info_request'            : function(obj) {  const buffer = getStakeinfo(obj.stake_info_request); return buffer},
+  'avg_reward_request'            : function(obj) {  const buffer = getAvgReward(); return buffer},
+  'reward_penalty_request'        : function(obj) {  const buffer = getRewardPenalty(obj.reward_penalty_request); return buffer},
+  'apy_request'                  : function(obj) {  const buffer = getApy(obj.apy_request); return buffer},
+  'commission_request'           : function(obj) {  const buffer = getCommission(obj.commission_request); return buffer},
+};
+
+function getUnknowrequestBuff() {
```
Comment (jasl) [inline backend/index.js]: getUnknownRequestBuffer

**Review by jasl — CHANGES_REQUESTED** [review]

**Comment by jasl** [comment]
Unknow should be Unknown

File: backend/index.js [hunk 10]
```diff
@@ -1 +1,89 @@
...
+      // 监听end事件
+      ctx.req.addListener('end', function(error) {
+          if (error) {
+              console.log("error:", error)
+          } else {
+              const resultBuff = Buffer.concat(buffers) // 合并接收到的buffer数据
+              const result = protobuf.CommonRequest.decode(resultBuff); // 解码接受到的 buffer 数据
+              const obj = protobuf.CommonRequest.toObject(result)  // 转换成json对象
+
+              const firstProperity = Object.keys(obj)[0]
```
Comment (jasl) [inline backend/index.js]: Property

File: backend/index.js [hunk 11]
```diff
@@ -1 +1,89 @@
...
+  const message = protobuf.CommonResponse.create({status: {msg: "unknow request", success: -1}});
+  const buffer = protobuf.CommonResponse.encode(message).finish();
+  return buffer
+}
+
+// 解析上下文里node原生请求的POST参数
+function parsePostData( ctx ) {
+  return new Promise((resolve, reject) => {
+    try {
+      var buffers = [];
```
Comment (jasl) [inline backend/index.js]: let buffers

File: backend/lib/api/apy.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    .limit(30*24).lean();//30 days
+
+  if (!historyRoundInfo) {
+    const message = protobuf.ApyResponse.create({ status: { success: -1, msg: 'can not find data in database' } });
+    const buffer = protobuf.ApyResponse.encode(message).finish();
+    return buffer;
+  }
+
+  const filterWorkers = historyRoundInfo.map(roundInfo => roundInfo.workers)
+    .filter((element, index) => {return 0 == index % 4} )
```
Comment (jasl) [inline backend/lib/api/apy.js]: 0 === index % 4

File: backend/lib/api/commission.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    .limit(30*24).lean();//30 days
+
+  if (!historyRoundInfo) {
+    const message = protobuf.CommissionResponse.create({ status: { success: -1, msg: 'can not find data in database' } });
+    const buffer = protobuf.CommissionResponse.encode(message).finish();
+    return buffer;
+  }
+
+  const filterWorkers = historyRoundInfo.map(roundInfo => roundInfo.workers)
+    .filter((element, index) => {return 0 == index % 4} )
```
Comment (jasl) [inline backend/lib/api/commission.js]: 0 === index % 4

File: backend/servers/history.js [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+      const available_supply = tokeninfo.available_supply
+      if (0 === available_supply) {
+        return 0
+      }
+      
+      return stakeSum.div(available_supply)
+    }
+
+    const getApyCurrentRound = function(accumulatedFire2PHA, stakeSumOfUserStake) {
+      if (0 == stakeSumOfUserStake) {
```
Comment (jasl) [inline backend/servers/history.js]: ===

**Comment by jasl** [comment]
prefer to use `===`

**Review by jasl — CHANGES_REQUESTED** [review]

File: backend/servers/history.js (hunk 6 above)
Comment (baidang201) [reply backend/servers/history.js]: done

File: backend/servers/history.js (hunk 7 above)
Comment (baidang201) [reply backend/servers/history.js]: done

File: backend/servers/history.js (hunk 8 above)
Comment (baidang201) [reply backend/servers/history.js]: done

File: backend/index.js (hunk 9 above)
Comment (baidang201) [reply backend/index.js]: done

File: backend/index.js (hunk 10 above)
Comment (baidang201) [reply backend/index.js]: done

File: backend/index.js (hunk 11 above)
Comment (baidang201) [reply backend/index.js]: done

File: backend/lib/api/apy.js (hunk 12 above)
Comment (baidang201) [reply backend/lib/api/apy.js]: done

File: backend/lib/api/commission.js (hunk 13 above)
Comment (baidang201) [reply backend/lib/api/commission.js]: done

File: backend/servers/history.js (hunk 14 above)
Comment (baidang201) [reply backend/servers/history.js]: done

**Pull request merged**